src/CabanaPD_Force.hpp

```diff
@@ -139,19 +139,68 @@ class Force<ExecutionSpace, BaseForceModel>
     Timer _energy_timer;
 
   public:
-    Force( const bool half_neigh )
+    using neighbor_type =
+        Cabana::VerletList<MemorySpace, Cabana::FullNeighborTag,
+                           Cabana::VerletLayout2D, Cabana::TeamOpTag>;
+    neighbor_type _neigh_list;
+
+    // Primary constructor: use positions and construct neighbors.
+    template <class ParticleType>
+    Force( const bool half_neigh, const double delta,
+           const ParticleType& particles )
         : _half_neigh( half_neigh )
+        , _neigh_list( neighbor_type( particles.sliceReferencePosition(), 0,
+                                      particles.n_local, delta + 1e-14, 1.0,
+                                      particles.ghost_mesh_lo,
+                                      particles.ghost_mesh_hi ) )
     {
     }
 
-    template <class ParticleType, class NeighListType, class ParallelType>
-    void computeWeightedVolume( ParticleType&, const NeighListType&,
-                                const ParallelType ) const
+    // Primary constructor: use positions and construct neighbors.
+    template <class NeighborType>
+    Force( const bool half_neigh, const NeighborType& neighbors )
+        : _half_neigh( half_neigh )
+        , _neigh_list( neighbors )
+    {
+    }
+
+    auto getMaxLocalNeighbors()
     {
+        // NOTE: this only works on host for 2D Verlet!
```

Comment:
Use an assert here on `MemorySpace` and/or `NeighborListType` to enforce.

Reply:
Put in a kernel because we don't expose those member types..